ui/src/serverIO.js

```diff
@@ -104,16 +104,12 @@ class ServerIO {
 
     this.hwrSocket.on('disconnect', (reason) => {
       console.log('hwrSocket disconnected!'); // eslint-disable-line no-console
-      const socket = this.hwrSocket;
```

Comment:
This was the bug: we're saving a reference to the old socket that was disconnected. Since we recreate a socket every time we call `listen()`, `socket` would keep referring to the previous, disconnected socket, and `socket.connected` would remain `false`.